### PR TITLE
fix(donations): resolve often reported "Error: Nonce verification has failed." #3200

### LINKED ISSUE #3200: fix(donations): resolve often reported "Error: Nonce verification has failed."
## User Story

As a `donor`, I want `give` without seeing a `nonce verification` error.

## Current Behavior

This has been reported often on WP.org

https://wordpress.org/support/topic/error-nonce-verification-has-failed-4/

## Steps to Reproduce

I'm not sure at the moment. More info is needed. @mathetos do you know how we can reproduce this issue?

## Possible Solution

Caching issues.

### PR DESCRIPTION
**Pull request by ravinderk**
## Description
This pr will resolve #3200 

## How Has This Been Tested?
Tested with https://wordpress.org/plugins/wp-fastest-cache/

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

### CODE REVIEW

**Comment by DevinWalker** [comment]
@ravinderk this looks good but can you also clean up `give_donation_form_nonce()` with a `give_clean()` around the `$_POST` request for wpcs. And also take note how Caldera is doing it particularly the saving of an HTTP request if the nonce is not more than an hour old:

File: assets/src/js/frontend/give-ajax.js [hunk 1]
```diff
@@ -17,6 +17,12 @@ jQuery( document ).ready( function( $ ) {
 	//Hide loading elements
 	$( '.give-loading-text' ).hide();
 
+	// Create ,update and invalidate cached nonce.
+	jQuery('.give-form').each( function( index, $form ){
```
Comment (DevinWalker) [inline assets/src/js/frontend/give-ajax.js]: Why `jQuery` rather than `$` used above and below?

File: assets/src/js/frontend/give-ajax.js [hunk 2]
```diff
@@ -17,6 +17,12 @@ jQuery( document ).ready( function( $ ) {
 	//Hide loading elements
 	$( '.give-loading-text' ).hide();
 
+	// Create ,update and invalidate cached nonce.
```
Comment (DevinWalker) [inline assets/src/js/frontend/give-ajax.js]: Fix spacing typo `Create ,`

File: assets/src/js/frontend/give-donations.js [hunk 3]
```diff
@@ -604,6 +608,22 @@ Give.form = {
...
+		 * Check if nonce old or not
+		 *
+		 * @since 2.1.3
+		 * @param {object} $form
+		 *
+		 * @return {boolean}
+		 */
+		isNonceOld: function ($form) {
+			const $nonceField = jQuery('#_wpnonce', $form),
+				nonceTime = parseInt( $nonceField.data('time') ) + parseInt( $nonceField.data('nonce-life') ) - 600,
```
Comment (ravinderk) [inline assets/src/js/frontend/give-donations.js]: @DevinWalker Logic will only reset nonce when nonce reach to its closest possible expire date. I take `30 minute` as threshold time to process donation by donor after loading donation page.

**Comment by ravinderk** [comment]
@DevinWalker `give_clean`already added by @mehul0810 in `give_donation_form_nonce` in spam donor related pr. I will create pr today for that.

**Comment by ravinderk** [comment]
@DevinWalker There will not be any notice for the deprecated nonce and neither they will be allowed to process donation because of their deprecated nonce. They can process donation by reloading donation page.

**Comment by DevinWalker** [comment]
> There will not be any notice for the deprecated nonce and neither they will be allowed to process donation because of their deprecated nonce.

This doesn't sound like a good donor experience to me then. Please ensure the following notices appear:

**5 minutes prior to nonce expiration:**

> You have 5 minutes remaining to complete your donation. 

**Upon expiration**

> Your donation timed out, please refresh the page to donate.

**Comment by ravinderk** [comment]
@DevinWalker Added nonce expire status notices: https://github.com/WordImpress/Give/pull/3213#issuecomment-389043635



image

**Comment by ravinderk** [comment]
> Slack Chat Summary
> 
> Participants: @DevinWalker, Ravinder Kumar
> Topic: Do not show a message and performance improvement
> Result:  do not show a notice when `nonce` expires, reset nonce in the background and code refactored to improve js performance

**Pull request merged**